File: 2017/05/jump_maze.go
package main

import (
	"fmt"
	"io/ioutil"
	"strconv"
	"strings"
)

func readFile(filename string) (numbers []int) {
	dat, _ := ioutil.ReadFile("input")
	lines := strings.Split(string(dat), "\n")
	for _, line := range lines {
		n, _ := strconv.Atoi(line)
		numbers = append(numbers, n)
	}
	return
}

func cpu1(instructions []int) (jumps int) {
	var next int
	for i := 0; i < len(instructions) && i >= 0; {
		next += instructions[i]
		instructions[i]++
		jumps++
		i = next
	}
	return
}

func cpu2(instructions []int) (jumps int) {
	var next int
	for i := 0; i < len(instructions) && i >= 0; {
		next += instructions[i]
		if instructions[i] >= 3 {
			instructions[i]--
		} else {
			instructions[i]++
		}
		jumps++
		i = next
	}
	return
}

func main() {

	fmt.Println(cpu1(readFile("input")))
	fmt.Println(cpu2(readFile("input")))

}

File: 2017/05/jump_maze.py
##!/usr/bin/env python3

def getInstructions():
    with open("input") as f:
        return [int(line) for line in f.readlines()]

def cpu1(instructions):
    i = 0
    jumps = 0
    while i >= 0 and i < len(instructions):
        n = i + instructions[i]
        instructions[i] += 1
        i = n
        jumps += 1
    return jumps

def cpu2(instructions):
    i = 0
    jumps = 0
    while True:
        try:
            x = instructions[i]
            n = i + x
            if x > 2:
                instructions[i] -= 1
            else:
                instructions[i] += 1
            i = n
            jumps += 1
        except IndexError:
            break
    return jumps

if __name__ == "__main__":
    print(cpu1(getInstructions()))
    print(cpu2(getInstructions()))